include/channel.h

```diff
@@ -0,0 +1,203 @@
+// Channels are objects which enable iteration over other Channel objects with
+// the same name/identifier.
+// This implementation resembles a map, mapping identifiers to lists. 
+// Channels are implemented as an item of two separate lists.
+// The "unique" list holds items with different identifiers,
+// the "same" list holds items that share the same identifier. 
+// Each list item itself is a valid list containing only itself;
+// multiple items can be combined to a larger list.
+// 
+// This implementation constructs a single valid "unique" list,
+// starting with a globally defined root node. All "same" lists of all items
+// are considered valid.
+//
+// Adding to this structure implies iterating the "unique" list, until either
+// the same identifier is found (and the new item is added to the "same" list
+// of the found item) or the end of the "unique" list reached (and it's
+// appended as new "unique" list item).
+//
+// Each added Channel object is now part of a "same" list which can be directly
+// iterated to find other items.
+//
+// exemplary structure:
+// 
+//              root node---.  O        O
+//                          v  |        |
+// unique identifier list-> O--O--O--O--O  
+//                             |     |
+//                             O     O
+//                             ^
+//      same identifier list---'                    
+//
+// Each Channel object contains a context pointer, a callback function and
+// an additional flags field. A broadcast function is provided to iterate
+// through all "same" list items of a Channel object. The callback is issued
+// with a data pointer and a timeout value by the caller and the callee
+// specific context and flags. This interface is intended to enable dynamic
+// management of FreeRTOS queues, where a queue can be registered as a context
+// of a Channel and therefore fed by the Channels callback.
+
+#ifndef __CHANNEL__
+#define __CHANNEL__
+
+#include <freertos/FreeRTOS.h>
+#include <freertos/queue.h>
+
+#include "list.h"
+
+typedef BaseType_t (*Channel_callback) (void *ctx, const void * const data, TickType_t timeout, const BaseType_t flags);
+
+typedef struct channel {
+    const char *identifier;
+    struct list_head same;
+    struct list_head unique;
+    void *ctx;
+    BaseType_t flags;
+    Channel_callback callback;
+} Channel;
```

Comment:
Why is channel uppercase here? Shouldn't it be lowercase as the rest of typedefs?
```suggestion
} channel;
```

Reply:
this depends on our coding guidelines, should complex types start with lower or uppercase?